pkg/provision/cleanup.go

```diff
@@ -0,0 +1,234 @@
+		}
+
+		qgroups, err := utils.QGroupList(ctx, path)
+		if err != nil {
+			return errors.Wrapf(err, "failed to list available cgroups in: %s", path)
+		}
+
+		for _, subvol := range subvols {
+			log.Info().Msgf("checking subvol %s", subvol.Path)
+			// Don't delete zos-cache!
+			if subvol.Path == storage.CacheLabel {
+				continue
+			}
+
+			qgroup, ok := qgroups[fmt.Sprintf("0/%d", subvol.ID)]
+			if !ok {
+				log.Info().Msgf("skipping volume '%s' has no assigned quota", subvol.Path)
+				continue
+			}
+
+			// Now, is this subvol in one of the toSave ?
+			if _, ok := toSave[filepath.Base(subvol.Path)]; ok {
+				log.Info().Msgf("skipping volume '%s' is used", subvol.Path)
+				continue
+			}
+
+			// Now, is this subvol in one of the toDelete ?
+			if _, ok := toDelete[filepath.Base(subvol.Path)]; ok {
+				// delete the subvolume
+				delete := checkReservationToDelete(subvol.Path, client)
+				if delete {
+					log.Info().Msgf("deleting subvolume %s", subvol.Path)
+					if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+						log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+					}
+					if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+						log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+					}
+					continue
+				}
+				log.Info().Msgf("skipping subvolume %s", subvol.Path)
+				continue
+			}
+
+			// Is this subvol not in toDelete and not in toSave?
+			// Check the explorer if it needs to be deleted
+			delete := checkReservationToDelete(subvol.Path, client)
+			if delete {
+				log.Info().Msgf("deleting subvolume %s", subvol.Path)
+				if err := utils.SubvolumeRemove(ctx, filepath.Join(path, subvol.Path)); err != nil {
+					log.Err(err).Msgf("failed to delete subvol '%s'", subvol.Path)
+				}
+				if err := utils.QGroupDestroy(ctx, qgroup.ID, path); err != nil {
+					log.Err(err).Msgf("failed to delete qgroup: '%s'", qgroup.ID)
+				}
+				continue
+			}
+			log.Info().Msgf("skipping subvolume %s", subvol.Path)
+		}
+	}
+
+	return nil
+}
+
+func checkReservationToDelete(path string, cl *client.Client) bool {
+	log.Info().Msgf("checking explorer for reservation: %s", path)
+	reservation, err := cl.Workloads.NodeWorkloadGet(path)
+	if err != nil {
+		var hErr client.HTTPError
+		if ok := errors.As(err, &hErr); ok {
+			resp := hErr.Response()
+			// If reservation is not found it should be deleted
+			if resp.StatusCode == 404 {
+				return true
+			}
+		}
+		return false
+	}
+
+	if reservation.GetNextAction() == workloads.NextActionDelete {
+		log.Info().Msgf("subvolume with path %s has next action to delete", path)
+		return true
+	}
+
+	return false
+}
+
+// checks running containers for subvolumes that might need to be saved because they are used
+// and subvolumes that might need to be deleted because they have no attached container anymore
+func checkContainers(msgBrokerCon string) (map[string]struct{}, map[string]struct{}, error) {
```

Comment:
I would instead pass the connection to zbus and not the connection url. Hence i don't have to create a new connection pool every time. Possible leakage.